docs/source/cpp/env_vars.rst

```diff
@@ -0,0 +1,128 @@
+   :ref:`building Arrow C++ <building-arrow-cpp>`.
+
+.. envvar:: ARROW_LIBHDFS_DIR
+
+   The directory containing the C HDFS library (``hdfs.dll`` on Windows,
+   ``libhdfs.dylib`` on macOS, ``libhdfs.so`` on other platforms).
+   Alternatively, one can set :envvar:`HADOOP_HOME`.
+
+.. envvar:: ARROW_TRACING_BACKEND
+
+   The backend where to export `OpenTelemetry <https://opentelemetry.io/>`_-based
+   execution traces.  Possible values are:
+
+   - ``ostream``: emit textual log messages to stdout;
+   - ``otlp_http``: emit JSON traces to a HTTP server (by default, the endpoint
+     URL is "http://localhost:4318/v1/traces");
+   - ``arrow_otlp_stdout``: emit JSON traces to stdout;
+   - ``arrow_otlp_stderr``: emit JSON traces to stderr.
+
+   This environment variable has no effect if Arrow C++ was not built with
+   tracing enabled.
+
+   .. seealso::
+
+      `OpenTelemetry configuration for remote endpoints
+      <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md>`__
+
+.. envvar:: ARROW_USER_SIMD_LEVEL
+
+   The SIMD optimization level to select.  By default, Arrow C++ detects
+   the capabilities of the current CPU at runtime and chooses the best
+   execution paths based on that information.  One can override the detection
+   by setting this environment variable to a well-defined value.
+   Supported values are:
+
+   - ``NONE`` disables any runtime-selected SIMD optimization;
+   - ``SSE4.2`` enables any SSE2-based optimizations until SSE4.2 (included);
+   - ``AVX`` enables any AVX-based optimizations and earlier;
+   - ``AVX2`` enables any AVX2-based optimizations and earlier;
+   - ``AVX512`` enables any AVX512-based optimizations and earlier.
+
+   This environment variable only has an effect on x86 platforms.  Other
+   platforms currently do not implement any form of runtime dispatch.
+
+   .. note::
+      In addition to runtime dispatch, the compile-time SIMD level can
+      be set using the ``ARROW_SIMD_LEVEL`` CMake configuration variable.
+      Unlike runtime dispatch, compile-time SIMD optimizations cannot be
+      changed at runtime (for example, if you compile Arrow C++ with AVX512
+      enabled, the resulting binary will only run on AVX512-enabled CPUs).
+
+.. envvar:: GANDIVA_CACHE_SIZE
+
+   The number of entries to keep in the Gandiva JIT compilation cache.
+   The cache is in-memory and does not persist accross processes.
+
+.. envvar:: HADOOP_HOME
+
+   The path to the Hadoop installation.
+
+.. envvar:: JAVA_HOME
+
+   The path to the Java Runtime Environment installation.  This may be
+   required for HDFS support if Java is installed in a non-standard location.
+
+.. envvar:: OMP_NUM_THREADS
+
+   The number of worker threads in the global (process-wide) CPU thread pool.
+   If this environment variable is not defined, the available hardware
+   concurrency is determined using a platform-specific routine.
```

Comment:
@westonpace I notice the IO thread pool size cannot be influenced for now, unless I'm mistaken. Is this something we'd like to make configurable?

Reply:
There is a function, though not an env var: https://github.com/apache/arrow/blob/5cb5afc40547b4f75739e31ff8632c71a10d3084/cpp/src/arrow/io/type_fwd.h#L46-L52

Also, hmm. https://github.com/apache/arrow/blob/5cb5afc40547b4f75739e31ff8632c71a10d3084/r/src/threadpool.cpp#L51-L57

Reply:
:-D. Do you want to open a JIRA for ?

Reply:
At the moment I think it is very important this is configurable.  So I am +1 on being able to expose this via an environment variable.  We have had at least one customer that was using S3 and benefited from setting this larger than the initial default.

At some point though I think we want to move towards having I/O context / thread pools specific to the filesystem.  A single global default doesn't make a lot of sense when you might have a mix of local and remote workloads.  Even then I suppose we might still have a global default as a fallback in case the user doesn't specify anything.

Reply:
Ok, I created https://issues.apache.org/jira/browse/ARROW-15941 for an IO thread pool environment variable.